distributor.go

```diff
@@ -59,7 +60,8 @@ type Distributor struct {
 type ReadRing interface {
 	prometheus.Collector
 
-	Get(key uint32, n int) ([]ring.IngesterDesc, error)
+	Get(key uint32, n int, op ring.Operation) ([]ring.IngesterDesc, error)
+	BatchGet(keys []uint32, n int, op ring.Operation) ([][]ring.IngesterDesc, error)
```

Comment:
Yes we're always reading from the ring, but the operation says what that info is for - so we read from the ring when writing samples and when processing queries.  The way we handle leaving node depends on why we're asking...